### PR TITLE
Patch redis-py bug for Windows

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

Socket disconnection result in errors on Windows rather than empty reads, which redis-py does not handle. This patch handles translates connection resets to empty reads.

## Related issue number

#631

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/compat.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import errno
+import socket
+import sys
+
+
+def patch_redis():
```
Comment (robertnishihara) [inline python/ray/compat.py:6]: If we need to do this, then there should be a detailed comment explaining the issue (or a link to the relevant github issue).

**Review by edoakes — DISMISSED** [review]
Please add some comments/a link to a github issue in the code for why this is needed. Also, given that this is just one specific patch, I think it would be better to rename `patch_redis` to be more specific (e.g., `patch_redis_empty_recv` and then call that from `__init__.py` instead of `apply_patches()`. If we need more patches in the future, which seems unlikely, we can always make it general.

File: python/ray/__init__.py [hunk 2]
```diff
@@ -30,6 +30,9 @@
     os.path.abspath(os.path.dirname(__file__)), "thirdparty_files")
 sys.path.insert(0, thirdparty_files)
 
+import ray.compat  # noqa: E402
+ray.compat.patch_redis_empty_recv()
```
Comment (edoakes) [inline python/ray/__init__.py:34]: Can we actually put the windows check here? And then an assert inside of the patch function. That would make it clear both from here that we're only patching windows and from in the function that it should only be called for windows.

Comment (mehrdadn) [reply python/ray/__init__.py]: Okay, just did.

**Review by edoakes — APPROVED** [review]

**Pull request merged**